MIT_6.S081/2020/EXAMPLE_PROCESS.md

```diff
@@ -0,0 +1,178 @@
+# XV6 Process Creation Tutorial
+
+Assuming you have successfully installed xv6 and you when you run `make qemu`, you get to this console output:
+```
+xv6 kernel is booting
+
+hart 1 starting
+hart 2 starting
+init: starting sh
+$
+```
+
+This tutorial demonstrates how to create and run a simple process example from Chapter 1 of the MIT XV6 RISC-V [Book](https://pdos.csail.mit.edu/6.828/2020/xv6/book-riscv-rev1.pdf).
+
+We'll create a parent-child process relationship and explore basic process management in XV6.
+
+## Prerequisites
+
+- XV6 operating system source code
+- RISC-V toolchain installed
+- Basic understanding of C programming
+
+## Creating the Example
+
+### 1. File Setup
+
+Create a new file `example_process.c` in the user directory of your xv6-labs-2020 repo:
+
+```bash
+touch user/example_process.c
+```
+
+### 2. Source Code
+
+Add the following code to `example_process.c`:
+
+```c
+// Copyright 2024 <your-name>
+// SPDX-License-Identifier: MIT
+
+#include "kernel/types.h"
+#include "kernel/stat.h"
```

Comment:
nit: we don't need these two headers for our simple program that's calling only `fork`, `exit` and `printf`:
```
"kernel/stat.h"
"kernel/fs.h"
```
We can remove those, plus the respective explanations.

However, good stuff on all these!

Reply:
It doesn't hurt to explain that we are using the xv6 interface here, a similar program on Linux for example will have the following headers:

```c
#include <unistd.h>    // For fork()
#include <sys/types.h> // For pid_t
#include <stdio.h>     // For printf(), perror()
#include <stdlib.h>    // For exit()

int main() {
    pid_t pid = fork(); // Create a new process

    if (pid < 0) {
        // Error occurred
        perror("fork failed");
        exit(EXIT_FAILURE);
    } else if (pid == 0) {
        // Child process
        printf("Hello from the child process! PID: %d\n", getpid());
    } else {
        // Parent process
        printf("Hello from the parent process! PID: %d, Child PID: %d\n", getpid(), pid);
    }

    return 0;
}

```

Reply:
done